File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.cpp
class Solution {
public:
    int minFlips(string s) {
        int res = 0;
        vector<vector<int>> freqs(2, vector<int>(2, 0));
        for(int i = 0; i < s.length(); i++)
        {
            if(i % 2 == 0)
            {
                freqs[0][s[i] - '0']++;
            }
            else
            {
                freqs[1][s[i] - '0']++;
            }
        }
        if(s.length() % 2 == 0)
        {
            return s.length() - max(freqs[0][0] + freqs[1][1], freqs[0][1] + freqs[1][0]);
        }
        for(int i = 0; i < s.length(); i++)
        {
            swap(freqs[0][0], freqs[1][0]);
            swap(freqs[0][1], freqs[1][1]);
            if(s[i] == '0')
            {
                freqs[0][0]++;
                freqs[1][0]--;
            }
            else
            {
                freqs[0][1]++;
                freqs[1][1]--;
            }
            res = max(res, max(freqs[0][0] + freqs[1][1], freqs[0][1] + freqs[1][0]));
        }
        return s.length() - res;
    }
};

File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.py
class Solution:
    def minFlips(self, s: str) -> int:
        res = 0
        freqs = [[0, 0], [0, 0]]
        for i in range(len(s)):
            if i % 2 == 0:
                freqs[0][int(s[i])] += 1
            else:
                freqs[1][int(s[i])] += 1
        if len(s) % 2 == 0:
            return len(s) - max(freqs[0][0] + freqs[1][1], freqs[0][1] + freqs[1][0])
        for i in range(len(s)):
            freqs[0][0], freqs[1][0] = freqs[1][0], freqs[0][0]
            freqs[0][1], freqs[1][1] = freqs[1][1], freqs[0][1]
            if s[i] == '0':
                freqs[0][0] += 1
                freqs[1][0] -= 1
            else:
                freqs[0][1] += 1
                freqs[1][1] -= 1
            res = max(res, freqs[0][0] + freqs[1][1], freqs[0][1] + freqs[1][0])
        return len(s) - res